common/scala/src/main/scala/org/apache/openwhisk/core/entity/ParameterEncryption.scala

```diff
@@ -0,0 +1,163 @@
+        })))
+  }
+  def unlock(params: Parameters): Parameters = {
+    val configuredEncryptors = new encrypters(storageConfig)
+    new Parameters(
+      params.getMap
+        .map(({
+          case (paramName, paramValue)
+              if paramValue.encryption != JsNull && !configuredEncryptors
+                .getEncrypter(paramValue.encryption.convertTo[String])
+                .isEmpty =>
+            paramName -> configuredEncryptors
+              .getEncrypter(paramValue.encryption.convertTo[String])
+              .get
+              .decrypt(paramValue)
+          case (paramName, paramValue) => paramName -> paramValue
+        })))
+  }
+}
+
+private trait encrypter {
+  def encrypt(p: ParameterValue): ParameterValue
+  def decrypt(p: ParameterValue): ParameterValue
+  val name: String
+}
+
+private class encrypters(val storageConfig: ParameterStorageConfig) {
+  private val availableEncrypters = Map("" -> new NoopCrypt()) ++
+    (if (!storageConfig.aes256.isEmpty) Some(Aes256.name -> new Aes256(getKeyBytes(storageConfig.aes256))) else None) ++
+    (if (!storageConfig.aes128.isEmpty) Some(Aes128.name -> new Aes128(getKeyBytes(storageConfig.aes128))) else None)
+
+  protected[entity] def getCurrentEncrypter(): encrypter = {
+    availableEncrypters.get(ParameterEncryption.storageConfig.current).get
+  }
+  protected[entity] def getEncrypter(name: String) = {
+    availableEncrypters.get(name)
+  }
+
+  def getKeyBytes(key: String): Array[Byte] = {
+    if (key.length == 0) {
+      Array[Byte]()
+    } else {
+      Base64.getDecoder().decode(key)
+    }
+  }
+}
+
+private trait AesEncryption extends encrypter {
+  val key: Array[Byte]
+  val ivLen: Int
+  val name: String
+  private val tLen = 128
+  private val secretKey = new SecretKeySpec(key, "AES")
+
+  private val secureRandom = new SecureRandom()
+
+  def encrypt(value: ParameterValue): ParameterValue = {
+    val iv = new Array[Byte](ivLen)
+    secureRandom.nextBytes(iv)
+    val gcmSpec = new GCMParameterSpec(tLen, iv)
+    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
+    cipher.init(Cipher.ENCRYPT_MODE, secretKey, gcmSpec)
+    val clearText = value.value.compactPrint.getBytes(StandardCharsets.UTF_8)
+    val cipherText = cipher.doFinal(clearText)
+
+    val byteBuffer = ByteBuffer.allocate(4 + iv.length + cipherText.length)
+    byteBuffer.putInt(iv.length)
+    byteBuffer.put(iv)
+    byteBuffer.put(cipherText)
+    val cipherMessage = byteBuffer.array
+    ParameterValue(JsString(Base64.getEncoder.encodeToString(cipherMessage)), value.init, Some(JsString(name)))
+  }
+
+  def decrypt(value: ParameterValue): ParameterValue = {
+    val cipherMessage = value.value.convertTo[String].getBytes(StandardCharsets.UTF_8)
+    val byteBuffer = ByteBuffer.wrap(Base64.getDecoder.decode(cipherMessage))
+    val ivLength = byteBuffer.getInt
+    if (ivLength != ivLen) {
+      throw new IllegalArgumentException("invalid iv length")
```

Comment:
decryption occurs in the invoker - this will therefore fail in the container proxy and we should add a test to cover failure and make sure the activation cleans up cleanly.